text/0000-pack-package-refactor.md

```diff
@@ -0,0 +1,58 @@
+# Meta
+[meta]: #meta
+- Name: (fill in the feature name: My Feature)
+- Start Date: 02/24/21
+- Author(s): dwillist
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a restructuring of the `pack` codebase into reusable modular packages.
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+    - Maintaining a modular and re-usable codebase enables contributors to be more effective, and encourages new adopters. 
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC is looking accomplish the following:
+- enable pack library consumers by moving code out of `internal`
+- ease the process of contribution by having well defined connectable components.
+- Follow a general process of if data is externally visible on a pack produced artifact. We should provide structs for decoding.
+
+To accomplish the above we propose the following package structure:
+
+```
+.
+├── cmd/
+│   ├── pack/ → current pack cli
+│   └── fetcher/ → cli to download buildpacks
+├── pkg/
+│   ├── archive/ → lib for converting multiple sources into OCI layers
+│   ├── client/ → current pack client library
+│   ├── descriptor/ → lib for processing project descriptor
+│   ├── inspect/ → lib for inspecting various CNB resources (app image, buildpack, builder, stack, ...)
+│   ├── internal/ → all non-exported go code
+│   ├── stack/ → lib for mixin validation & stack labels
+│   ├── stack/internal/ → keep internal details close to exported APIs.
```

Comment:
Given that whatever would live here could only be used by the `stack` package, should we just include it in the `stack` package but not export it?

Also, this package would be named `internal` which is a bit strange. If we really need an internal package (the logic is used by multiple exported packages) maybe we want `pkg/internal/X` where X describes the whatever utility it is exposing. I think it would be easier to maintain a useable and well organized public interface for `stacks` etc. if any stack specific functions had to go in the external package and internal packages were reserved for coherent but non-domain-level concepts (e.g. `blob`)